policy/daemon.go

```diff
@@ -317,11 +316,28 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
 		glg.Warn(errors.Wrap(err, "error body.close"))
 	}
 
-	// set eTag cache
-	eTag := res.Header.Get("ETag")
-	if eTag != "" {
-		glg.Debugf("Setting ETag %v for domain %s", eTag, domain)
-		p.etagCache.SetWithExpire(domain, &etagCache{eTag, sp}, p.etagExpTime)
+	// verify policy expiry
+	var etagValidDur time.Duration
+	if sp.SignedPolicyData.Expires == nil {
+		glg.Warnf("Warning: policy expiry not set, domain: %v", domain)
+		etagValidDur = p.etagExpTime - p.expireMargin
+	} else {
+		// if not set, sp.SignedPolicyData.Expires is nil
+		// if set, but invalid, sp.SignedPolicyData.Expires is Time{}
+		validDur := sp.SignedPolicyData.Expires.Time.Sub(fastime.Now())
+		if validDur <= p.expireMargin {
+			p.etagCache.Delete(domain)
+			glg.Errorf("Error verifying policy expiry, policy expiry: %v, valid duration: %v, valid margin: %v", sp.SignedPolicyData.Expires.Time, validDur, p.expireMargin)
+			return nil, false, errors.New("policy already expired")
+		}
+		etagValidDur = sp.SignedPolicyData.Expires.Time.Sub(fastime.Now()) - p.expireMargin
+	}
+
+	// set etag cache
+	etag := res.Header.Get("ETag")
+	if etag != "" {
+		glg.Debugf("Set domain %s with etag %v, duration: %s", domain, etag, etagValidDur)
```

Comment:
thinking we need to check etagValidDur <=0 before set to the cache.

Reply:
will fix